### ISSUE TITLE
Copy(somedir, Glob('otherdir/*')) fails because src list is dropped

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#3009]
This issue was originally created at: 2015-07-27 06:17:29.
This issue was reported by: `knut3`.
knut3 said at 2015-07-27 06:17:29
>Hi,
> 
I want to copy all files from a directory to another in a builder. I use:
 
```py
Execute(Copy('somedir', Glob('otherdir/*')))
```
 
This fails with error: `No such file or directory`
 
`Copy()` is implemented like this:
 
```py
def copy_func(dest, src):
    SCons.Node.FS.invalidate_node_memos(dest)
    if SCons.Util.is_List(src) and os.path.isdir(dest):
        for file in src:
            shutil.copy2(file, dest)
        return 0
    elif os.path.isfile(src):
        return shutil.copy2(src, dest)
    else:
        return shutil.copytree(src, dest, 1)
 
Copy = ActionFactory(copy_func,
                     lambda dest, src: 'Copy("%s", "%s")' % (dest, src),
                     convert=str)
```
 
---
 
I added some prints to see what's happening here:
 
```py
SCons.Util.is_List(src) => false
type(src) => str
```
 
`src` is converted to str by mistake. Thus, the wrong if clause is taken in `copy_func()`. This is caused by `convert=str` of the `ActionFactory`. If I remove this, scons fails somewhere else, but the list is preserved correctly.

**Comment by mwichmann** [comment]
This seems a valid complaint, though - the code at the head of the function unconditionally converts the two args, then proceeds to do checks based on the type - these two lines, which don't seem to be in the version the original poster commented on:
```py
    dest = str(dest)
    src = str(src)
...
    if SCons.Util.is_List(src) and os.path.isdir(dest):
```

**Comment by mwichmann** [comment]
My previous comment is hard to parse... anyway, just arrived at this issue by other means and did a "git blame" on the Copy code, and we come down to commit 5e3450b055 - which added the unconditional string conversions above as part of adding symlink support. That replaces the earlier `convert=str` mentioned in the original report. As a result it is impossible for the src argument to be a list, which is unusual for SCons syntax in general, but actually matches the extremely limited documentation (there's no docstring, only the manpage entry):

> Returns an Action object that will copy the `src` source file or directory to the `dest` destination file or directory.

So, if taken literally, "file or directory" would suggest you can't pass a list. Do we believe that?

**Comment by mwichmann** [comment]
Seems like a very small patch can fix this, if we decide it is *not* working as intended - which I now think it is not.
```py
index 65bd75ac2..274e9eb34 100644
--- a/SCons/Defaults.py
+++ b/SCons/Defaults.py
@@ -251,7 +251,10 @@ def copy_func(dest, src, symlinks=True):
     """
 
     dest = str(dest)
-    src = str(src)
+    if SCons.Util.is_List(src):
+        src = [str(n) for n in src]
+    else:
+        src = str(src)
 
     SCons.Node.FS.invalidate_node_memos(dest)
     if SCons.Util.is_List(src) and os.path.isdir(dest):
```

The reason I think this is just an oversight is just a few lines later there's an `is_List` check, and it does copies in a loop if so. @bdbaddog - should I generate a PR?

**Comment by mwichmann** [comment]
Additional note: the original example would be better written as: 
```py
Execute(Copy('somedir', Glob('otherdir/*', strings=True)))
```
because while a conversion from list-of-Nodes can take place inside the function called by the builder, this happens after the string function prints the command - so the output will look garbled (does not affect operation)

**Comment by mwichmann** [comment]
... abd that's easily fixable, Copy doesn't use a strfunc while all the related action-thingies do.

**Comment by bdbaddog** [comment]
not sure it's a valid complaint.
Using `Execute()` in a builder? 
That won't happen at builder/tree walk time, but rather immediately?

Yes a PR would be great.
(While the usage in the initial comment seems flawed, that the src arg can't be a list seems an error as well).
